File: android/app/src/main/java/com/strive/app/scanner/TomTomService.kt
package com.strive.scanner

import android.util.Log
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Appels TomTom natifs (geocoding + routing). Tourne dans le foreground service
 * donc non throttlé quand Strive est en background pendant un scan Uber/Bolt/Heetch.
 *
 * Utilise HttpURLConnection + Thread (cohérent avec GeminiVisionService).
 */
object TomTomService {

    private const val BASE_SEARCH  = "https://api.tomtom.com/search/2/geocode"
    private const val BASE_ROUTING = "https://api.tomtom.com/routing/1/calculateRoute"
    private const val COUNTRY_SET  = "FR,BE,CH,LU,GB,DE,ES,IT,NL,PT,AT,IE,PL"
    private const val TIMEOUT_MS   = 4_000
    private const val TAG          = "TomTomService"

    /** Clé API TomTom — renseignée depuis JS au démarrage via setTomTomApiKey. */
    var apiKey: String = ""

    val isReady get() = apiKey.isNotEmpty()

    data class Coords(val lat: Double, val lon: Double)
    data class GeocodeHit(val coords: Coords, val score: Double)
    data class RouteResult(val distanceKm: Double, val durationMin: Int)

    // Seuil de confiance sous lequel on retente avec des variantes d'adresse
    // (suffixe pays enlevé, seul code postal, etc.)
    private const val MIN_SCORE = 3.0

    /**
     * Calcule le trajet (distance + durée trafic) entre deux adresses texte.
     * Appelle le callback avec le résultat ou null si TomTom échoue.
     */
    fun calculateRoute(
        pickupAddress: String,
        destinationAddress: String,
        callback: (RouteResult?) -> Unit,
    ) {
        if (!isReady) { callback(null); return }
        if (pickupAddress.isBlank() || destinationAddress.isBlank()) { callback(null); return }

        Thread {
            try {
                // Geocoding pickup + destination en parallèle — chacun fait jusqu'à
                // 3 variantes d'adresse en série, donc parallèliser les 2 colonnes
                // divise potentiellement par 2 le temps total avant routing.
                var fromHit: GeocodeHit? = null
                var toHit: GeocodeHit? = null
                val tFrom = Thread { fromHit = geocodeBestVariant(pickupAddress) }
                val tTo = Thread { toHit = geocodeBestVariant(destinationAddress) }
                tFrom.start(); tTo.start()
                tFrom.join(); tTo.join()

                val from = fromHit
                val to = toHit
                if (from == null || to == null) {
                    Log.w(TAG, "geocode failed pickup=$from dest=$to")
                    callback(null); return@Thread
                }
                Log.i(TAG, "geocode scores pickup=${"%.1f".format(from.score)} dest=${"%.1f".format(to.score)}")
                val route = getRoute(from.coords, to.coords)
                callback(route)
            } catch (e: Exception) {
                Log.w(TAG, "exception", e)
                callback(null)
            }
        }.start()
    }

    // ─── Geocoding ────────────────────────────────────────────────────────────────

    /**
     * Essaie l'adresse telle quelle puis des variantes si le score est bas.
     * Garde le meilleur score. Early-exit si on trouve un résultat fiable.
     */
    private fun geocodeBestVariant(address: String): GeocodeHit? {
        val variants = buildAddressVariants(address)
        var best: GeocodeHit? = null
        for (variant in variants) {
            val hit = geocode(variant) ?: continue
            if (best == null || hit.score > best.score) best = hit
            if (best.score >= MIN_SCORE + 2) return best
        }
        // Plancher de fiabilité : un score < MIN_SCORE = match centroïde ville/pays
        // (adresse OCR douteuse). On préfère null → pas de "vraie" distance fausse :
        // le pipeline retombe sur l'OCR plutôt qu'un trajet centre-à-centre.
        return best?.takeIf { it.score >= MIN_SCORE }
    }

    /**
     * Variantes pour récupérer quand le géocodage de base échoue :
     * - texte brut
     * - sans suffixe ", France" / ", Belgique" (redondant, pollue parfois)
     * - code postal + ville seulement (sur les adresses trop précises qui matchent rien)
     */
    private fun buildAddressVariants(address: String): List<String> {
        val result = linkedSetOf<String>()
        val trimmed = address.trim()
        result.add(trimmed)
        // Strip trailing ", France/Belgique/Suisse/Luxembourg"
        val stripped = trimmed.replace(
            Regex(""",\s*(France|Belgique|Suisse|Luxembourg|Deutschland|España|Italia|Portugal|Ireland|Nederland)\s*$""", RegexOption.IGNORE_CASE),
            "",
        )
        if (stripped != trimmed) result.add(stripped)
        // Postal code + ville (si présent)
        val postalMatch = Regex("""\b(\d{4,5})\s+([A-Za-zÀ-ÿ][\w\s-]{2,40})""").find(trimmed)
        if (postalMatch != null) {
            result.add("${postalMatch.groupValues[1]} ${postalMatch.groupValues[2].trim()}")
        }
        return result.toList()
    }

    private fun geocode(address: String): GeocodeHit? {
        // Cache local — les coords GPS d'une adresse sont stables dans le temps.
        // Hit cache → 0 requête TomTom. Voir GeocodeCache pour la normalisation.
        GeocodeCache.get(address)?.let { return it }

        val encoded = URLEncoder.encode(address, "UTF-8")
        val url = "$BASE_SEARCH/$encoded.json?key=$apiKey&language=fr-FR&countrySet=$COUNTRY_SET&limit=1"
        val json = httpGet(url) ?: return null
        val hit = try {
            val results = JSONObject(json).optJSONArray("results") ?: return null
            if (results.length() == 0) return null
            val first = results.getJSONObject(0)
            val pos = first.getJSONObject("position")
            val score = first.optDouble("score", 0.0)
            GeocodeHit(Coords(pos.getDouble("lat"), pos.getDouble("lon")), score)
        } catch (e: Exception) {
            Log.w(TAG, "geocode parse", e); null
        }
        // Persiste uniquement les résultats fiables — un faux match (score
        // bas) cacherait à vie un mauvais POI. Le seuil MIN_SCORE est aussi
        // celui utilisé par geocodeBestVariant pour décider si on continue.
        if (hit != null && hit.score >= MIN_SCORE) GeocodeCache.put(address, hit)
        return hit
    }

    // ─── Routing ──────────────────────────────────────────────────────────────────

    private fun getRoute(from: Coords, to: Coords): RouteResult? {
        val waypoints = "${from.lat},${from.lon}:${to.lat},${to.lon}"
        // traffic=true + departAt=now → trafic temps réel (flux live + incidents)
        // à l'instant du calcul ; routeType=fastest → meilleur trajet.
        val url = "$BASE_ROUTING/$waypoints/json?key=$apiKey&travelMode=car&traffic=true&routeType=fastest&departAt=now"
        val json = httpGet(url) ?: return null
        return try {
            val summary = JSONObject(json)
                .getJSONArray("routes")
                .getJSONObject(0)
                .getJSONObject("summary")
            val seconds = summary.getInt("travelTimeInSeconds")
            val meters = summary.getInt("lengthInMeters")
            RouteResult(
                distanceKm = Math.round(meters / 100.0) / 10.0,
                durationMin = Math.round(seconds / 60.0).toInt(),
            )
        } catch (e: Exception) {
            Log.w(TAG, "route parse", e); null
        }
    }

    // ─── HTTP helper ──────────────────────────────────────────────────────────────

    private fun httpGet(url: String): String? {
        var conn: HttpURLConnection? = null
        return try {
            conn = (URL(url).openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
            }
            val code = conn.responseCode
            if (code != 200) { Log.w(TAG, "HTTP $code"); return null }
            conn.inputStream.bufferedReader().readText()
        } catch (e: Exception) {
            Log.w(TAG, "http", e); null
        } finally {
            conn?.disconnect()
        }
    }
}

File: ios/Strive/Scanner/TomTomService.swift
import Foundation

/// Service TomTom (geocoding + routing) — mirror exact du `TomTomService.kt`
/// côté Android. Utilisé par `AnalyzeRideIntent` (Shortcut iOS) et la Share
/// Extension pour vérifier la distance/durée annoncées par l'OCR.
///
/// Target Membership Xcode : ce fichier DOIT être inclus dans
///  - `Strive` (app principale)
///  - `StriveShareExtension`
///
/// Sinon le code ne sera pas accessible depuis la Share Extension.
final class TomTomService {

  static let shared = TomTomService()
  private init() {}

  private static let baseSearch  = "https://api.tomtom.com/search/2/geocode"
  private static let baseRouting = "https://api.tomtom.com/routing/1/calculateRoute"
  private static let countrySet  = "FR,BE,CH,LU,GB,DE,ES,IT,NL,PT,AT,IE,PL"
  private static let timeoutSec: TimeInterval = 4
  private static let minScore: Double = 3.0

  /// Stockée dans App Group par `ScanBridge.setTomTomApiKey`.
  private var cachedKey: String?

  private func apiKey() -> String? {
    if let k = cachedKey, !k.isEmpty { return k }
    let appGroupId = (Bundle.main.object(forInfoDictionaryKey: "StriveAppGroupId") as? String)
      ?? "group.com.striveapp.app"
    let k = UserDefaults(suiteName: appGroupId)?.string(forKey: "tomTomApiKey")
    cachedKey = k
    return (k?.isEmpty == false) ? k : nil
  }

  var isReady: Bool { apiKey() != nil }

  struct Coords { let lat: Double; let lon: Double }
  struct GeocodeHit { let coords: Coords; let score: Double }
  struct RouteResult { let distanceKm: Double; let durationMin: Int }

  /// Calcule le trajet entre 2 adresses texte. Renvoie nil si TomTom n'est
  /// pas configuré, si une adresse est vide, ou si geocoding/routing échoue.
  func calculateRoute(
    pickupAddress: String,
    destinationAddress: String,
    completion: @escaping (RouteResult?) -> Void
  ) {
    guard isReady else { completion(nil); return }
    let pickup = pickupAddress.trimmingCharacters(in: .whitespacesAndNewlines)
    let dest = destinationAddress.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !pickup.isEmpty, !dest.isEmpty else { completion(nil); return }

    DispatchQueue.global(qos: .userInitiated).async { [weak self] in
      guard let self = self else { completion(nil); return }
      // Geocoding pickup + destination en parallèle — divise par ~2 le temps
      // total. Chaque variant chain reste séquentiel à l'intérieur d'une
      // adresse (early-exit sur bon score), seuls les 2 fronts sont lancés.
      let group = DispatchGroup()
      var fromHit: GeocodeHit?
      var toHit: GeocodeHit?
      group.enter()
      DispatchQueue.global(qos: .userInitiated).async {
        fromHit = self.geocodeBestVariant(pickup)
        group.leave()
      }
      group.enter()
      DispatchQueue.global(qos: .userInitiated).async {
        toHit = self.geocodeBestVariant(dest)
        group.leave()
      }
      group.wait()

      guard let from = fromHit, let to = toHit else {
        completion(nil); return
      }
      let route = self.getRoute(from: from.coords, to: to.coords)
      DispatchQueue.main.async { completion(route) }
    }
  }

  // MARK: - Geocoding

  private func geocodeBestVariant(_ address: String) -> GeocodeHit? {
    var best: GeocodeHit?
    for variant in buildAddressVariants(address) {
      guard let hit = geocode(variant) else { continue }
      if best == nil || hit.score > best!.score { best = hit }
      if best!.score >= Self.minScore + 2 { return best }
    }
    return best
  }

  private func buildAddressVariants(_ address: String) -> [String] {
    var variants = [String]()
    var seen = Set<String>()
    func add(_ s: String) {
      let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !t.isEmpty, !seen.contains(t) else { return }
      seen.insert(t); variants.append(t)
    }

    let trimmed = address.trimmingCharacters(in: .whitespacesAndNewlines)
    add(trimmed)

    // Strip trailing ", France/Belgique/etc."
    let stripPattern = #",\s*(France|Belgique|Suisse|Luxembourg|Deutschland|España|Italia|Portugal|Ireland|Nederland)\s*$"#
    if let regex = try? NSRegularExpression(pattern: stripPattern, options: .caseInsensitive) {
      let range = NSRange(trimmed.startIndex..., in: trimmed)
      let stripped = regex.stringByReplacingMatches(in: trimmed, range: range, withTemplate: "")
      add(stripped)
    }

    // Postal + ville (5-digit postal usually)
    if let regex = try? NSRegularExpression(pattern: #"\b(\d{4,5})\s+([A-Za-zÀ-ÿ][\w\s\-]{2,40})"#),
       let match = regex.firstMatch(in: trimmed, range: NSRange(trimmed.startIndex..., in: trimmed)),
       match.numberOfRanges >= 3,
       let r1 = Range(match.range(at: 1), in: trimmed),
       let r2 = Range(match.range(at: 2), in: trimmed) {
      add("\(trimmed[r1]) \(trimmed[r2])")
    }
    return variants
  }

  private func geocode(_ address: String) -> GeocodeHit? {
    // Cache local — les coords GPS d'une adresse sont stables dans le temps.
    // Hit cache → 0 requête TomTom. Voir GeocodeCache pour la normalisation.
    if let cached = GeocodeCache.shared.get(address: address) { return cached }

    guard let key = apiKey(),
          let encoded = address.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)
    else { return nil }

    let urlString = "\(Self.baseSearch)/\(encoded).json?key=\(key)&language=fr-FR&countrySet=\(Self.countrySet)&limit=1"
    guard let url = URL(string: urlString),
          let data = httpGet(url),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let results = json["results"] as? [[String: Any]],
          let first = results.first,
          let pos = first["position"] as? [String: Any],
          let lat = (pos["lat"] as? NSNumber)?.doubleValue,
          let lon = (pos["lon"] as? NSNumber)?.doubleValue
    else { return nil }
    let score = (first["score"] as? NSNumber)?.doubleValue ?? 0
    let hit = GeocodeHit(coords: Coords(lat: lat, lon: lon), score: score)
    // Persiste uniquement les résultats fiables — un faux match (score bas)
    // cacherait à vie un mauvais POI. Seuil aligné sur geocodeBestVariant.
    if score >= Self.minScore { GeocodeCache.shared.put(address: address, hit: hit) }
    return hit
  }

  // MARK: - Routing

  private func getRoute(from: Coords, to: Coords) -> RouteResult? {
    guard let key = apiKey() else { return nil }
    let waypoints = "\(from.lat),\(from.lon):\(to.lat),\(to.lon)"
    // traffic=true + departAt=now → trafic temps réel (flux live + incidents) à
    // l'instant du calcul ; routeType=fastest → meilleur trajet.
    let urlString = "\(Self.baseRouting)/\(waypoints)/json?key=\(key)&travelMode=car&traffic=true&routeType=fastest&departAt=now"
    guard let url = URL(string: urlString),
          let data = httpGet(url),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let routes = json["routes"] as? [[String: Any]],
          let summary = routes.first?["summary"] as? [String: Any],
          let seconds = (summary["travelTimeInSeconds"] as? NSNumber)?.doubleValue,
          let meters = (summary["lengthInMeters"] as? NSNumber)?.doubleValue
    else { return nil }

    return RouteResult(
      distanceKm: (meters / 100.0).rounded() / 10.0,
      durationMin: Int((seconds / 60.0).rounded())
    )
  }

  // MARK: - HTTP helper

  private func httpGet(_ url: URL) -> Data? {
    var result: Data?
    let semaphore = DispatchSemaphore(value: 0)
    var req = URLRequest(url: url)
    req.timeoutInterval = Self.timeoutSec
    URLSession.shared.dataTask(with: req) { data, response, _ in
      defer { semaphore.signal() }
      guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return }
      result = data
    }.resume()
    _ = semaphore.wait(timeout: .now() + Self.timeoutSec + 1)
    return result
  }
}
